### PR TITLE
APP-1266: Add payment token

### PR DESCRIPTION
**Pull request by hugokallstrom**

### CODE REVIEW

File: app/src/main/java/com/hedvig/app/feature/offer/OfferViewModel.kt [hunk 1]
```diff
@@ -93,279 +84,152 @@ abstract class OfferViewModel : ViewModel() {
...
                         loginStatus = loginStatusService.getLoginStatus(),
-                        paymentMethods = getPaymentMethods()
+                        paymentMethods = offerModel.paymentApiResponseOrNull()
+                            ?: adyenRepository.paymentMethodsResponse(),
+                        externalProvider = it
                     )
                 }
-            )
+                .catch { _viewState.value = ViewState.Error(null) }
+                .launchIn(viewModelScope)
```
Comment (StylianosGakis) [inline app/src/main/java/com/hedvig/app/feature/offer/OfferViewModel.kt:200]: Hmm might have to be careful with a function called `toViewState` which feels like a mapping of a model to a viewState representation which is launching a never-ending coroutine. Especially considering that if we call `toViewState` again a second time, `launchIn` is called a second time with nothing cancelling the old collection (since `launchIn` does `scope.launch{ collect() }` under the hood). 
Kind of hard to follow how this all would function in such scenarios. Again don't have a clear fix for this, and I see that you had to refactor a pretty big/complex function so I can see how some of the features it had can be hard to maintain in this process.
The previous implementation was fixing this by having the collection be inside `transformLatest` or `collectLatest` which meant that previous collecting jobs would be cancelled when new data is incoming, but can't see how to apply the same thing now, at least not from the GitHub code view :D Kinda hard to follow everything

File: app/src/main/java/com/hedvig/app/feature/offer/ui/OfferActivity.kt [hunk 2]
```diff
@@ -363,7 +359,9 @@ class OfferActivity : BaseActivity(R.layout.activity_offer) {
                     model.onOpenCheckout()
                 }
             }
-            CheckoutMethod.APPROVE_ONLY -> model.approveOffer()
+            CheckoutMethod.APPROVE_ONLY -> lifecycleScope.launch {
+                model.approveOffer()
+            }
```
Comment (StylianosGakis) [inline app/src/main/java/com/hedvig/app/feature/offer/ui/OfferActivity.kt:364]: Are we sure we want to bind the job handling this to the view lifecycle coroutine scope instead of the viewmodel coroutine scope?
I mean like having the function not be suspending and instead launching a coroutine from the viewmodel instead like we usually do because it lives longer than the lifecycle one.

File: app/src/main/java/com/hedvig/app/util/apollo/PaymentMethodsApiResponseAdapter.kt [hunk 3]
```diff
@@ -3,13 +3,19 @@ package com.hedvig.app.util.apollo
 import com.adyen.checkout.components.model.PaymentMethodsApiResponse
 import com.apollographql.apollo.api.CustomTypeAdapter
 import com.apollographql.apollo.api.CustomTypeValue
+import com.hedvig.app.util.toJsonObject
 import org.json.JSONObject
 
 class PaymentMethodsApiResponseAdapter : CustomTypeAdapter<PaymentMethodsApiResponse> {
```
Comment (StylianosGakis) [inline app/src/main/java/com/hedvig/app/util/apollo/PaymentMethodsApiResponseAdapter.kt:9]: Maybe consider adding some comment on why this exists in the first place? The whole context on why we couldn't use a simple string and how this solves the problem we were facing? For future reference since otherwise this feels non-obvious why it exists without the context behind its creation.

**Review by StylianosGakis — COMMENTED** [review]
Hey quite some comments here. Sorry that a lot of them are just very local to what I see in specific places, pretty hard to follow the entire concept of all the things are changing here, so sometimes I focused on small chunks that I found that I could give some input on.

Aside from some more stylistic changes that you may/may not want to change like the `toEither{ Error(it) }` or using sealed interfaces instead of classes, I mostly saw things that may function slightly different which may be wrong compared to how they were before I think, for the non-quote cart cases mostly. 
Pretty hard to give feedback for the quote-cart specific things I think, but we can maybe consider the comments I had for the rest of the stuff and take if from there.

File: apollo/src/main/graphql/com/hedvig/android/owldroid/graphql/quoteCartFragment.graphql [hunk 4]
```diff
@@ -8,6 +8,17 @@ fragment QuoteCartFragment on QuoteCart {
...
   }
+  paymentConnection {
+    id
+    providers {
+      ... on Adyen {
+        availablePaymentMethods
+      }
+      ... on Trustly {
+        blank: _
+      }
```
[the remark this answers is not in the document]
Comment (hugokallstrom) [reply apollo/src/main/graphql/com/hedvig/android/owldroid/graphql/quoteCartFragment.graphql:19]: This type does not have any fields, but we need to query something so thats why we are using this "dummy" field.

File: app/src/main/java/com/hedvig/app/feature/adyen/ConnectPaymentUseCase.kt [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+class ConnectPaymentUseCase(
+    private val apolloClient: ApolloClient,
+    private val context: Context,
+    private val featureManager: FeatureManager,
+    private val marketManager: MarketManager,
+    private val graphQLQueryHandler: GraphQLQueryHandler,
+) {
+
+    sealed class Error {
```
[the remark this answers is not in the document]
Comment (hugokallstrom) [reply app/src/main/java/com/hedvig/app/feature/adyen/ConnectPaymentUseCase.kt:30]: This error is specifically used for ConnectPaymentUseCase, so we dont want to be able to inherit from it. I think a sealed class makes more sense for that, or am I missing something?

File: app/src/main/java/com/hedvig/app/feature/offer/OfferRepository.kt [hunk 6]
```diff
@@ -61,18 +51,47 @@ class OfferRepository(
         }
     }
 
+    suspend fun queryAndEmitOffer(quoteCartId: CreateQuoteCartUseCase.QuoteCartId?, quoteIds: List<String>) {
```
[the remark this answers is not in the document]
Comment (hugokallstrom) [reply app/src/main/java/com/hedvig/app/feature/offer/OfferRepository.kt:54]: Exactly, with this solution we need to call `queryAndEmitOffer` after every mutation done to the cart. This will cause `offerFlow` to re-emit. 

When we are calling .first() on the flow, we assume that there already is a value. Should we call `queryAndEmitOffer` in these places as well do you think? Or can we solve it another way?

File: app/src/main/java/com/hedvig/app/feature/offer/OfferRepository.kt [hunk 7]
```diff
@@ -61,18 +51,47 @@ class OfferRepository(
...
+        if (quoteCartId != null) {
+            val offer = queryQuoteCart(quoteCartId)
+            offerFlow.tryEmit(offer)
+        } else {
+            refreshOfferQuery(quoteIds)
```
[the remark this answers is not in the document]
Comment (hugokallstrom) [reply app/src/main/java/com/hedvig/app/feature/offer/OfferRepository.kt:59]: Good catch. Will fix

File: app/src/main/java/com/hedvig/app/feature/offer/ui/OfferActivity.kt (hunk 2 above)
Comment (hugokallstrom) [reply app/src/main/java/com/hedvig/app/feature/offer/ui/OfferActivity.kt:364]: Fixing!
